File: apps/android/app/src/main/java/site/smap/harubook/features/quiz/QuizViewModel.kt
package site.smap.harubook.features.quiz

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import site.smap.harubook.core.models.Quiz
import site.smap.harubook.core.models.QuizzesResponse
import site.smap.harubook.core.models.ReadingLogResponse
import site.smap.harubook.core.networking.ApiClient

/**
 * iOS `QuizViewModel.swift` 미러.
 *
 * 흐름:
 *  1. load() → POST /api/books/{id}/quiz 로 4지선다 5문제 생성/로드(서버 캐시).
 *  2. selectAnswer / next / prev 진행.
 *  3. submit() → reading_logs PATCH(quizScore + finishedAtUnix) + phase=Finished.
 */
class QuizViewModel(
    private val bookId: Int,
    private val readingLogId: Int?,
) : ViewModel() {

    sealed class Phase {
        data object Loading : Phase()
        data object Answering : Phase()
        data object Finished : Phase()
        data class Error(val message: String) : Phase()
    }

    data class UiState(
        val phase: Phase = Phase.Loading,
        val quizzes: List<Quiz> = emptyList(),
        val selections: Map<Int, Int> = emptyMap(),
        val currentIndex: Int = 0,
        val isSubmitting: Boolean = false,
    ) {
        val currentQuiz: Quiz? get() = quizzes.getOrNull(currentIndex)
        val totalQuestions: Int get() = quizzes.size
        val isLastQuestion: Boolean get() = currentIndex + 1 >= quizzes.size
        val score: Int get() = quizzes.count { selections[it.id] == it.answerIndex }
    }

    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    fun load() {
        viewModelScope.launch {
            _state.update { it.copy(phase = Phase.Loading) }
            try {
                val response: QuizzesResponse = ApiClient.post(path = "/api/books/$bookId/quiz")
                val sorted = response.quizzes.sortedBy { it.orderIndex }
                _state.update {
                    it.copy(
                        quizzes = sorted,
                        phase = if (sorted.isEmpty()) Phase.Error("퀴즈를 가져오지 못했습니다.") else Phase.Answering,
                    )
                }
            } catch (e: Throwable) {
                _state.update {
                    it.copy(phase = Phase.Error("퀴즈 로드 실패: ${e.message ?: e::class.java.simpleName}"))
                }
            }
        }
    }

    fun selectAnswer(choiceIndex: Int) {
        val quiz = _state.value.currentQuiz ?: return
        _state.update { it.copy(selections = it.selections + (quiz.id to choiceIndex)) }
    }

    fun goToNext() {
        _state.update { st ->
            if (st.currentIndex + 1 < st.quizzes.size) st.copy(currentIndex = st.currentIndex + 1) else st
        }
    }

    fun goToPrevious() {
        _state.update { st ->
            if (st.currentIndex > 0) st.copy(currentIndex = st.currentIndex - 1) else st
        }
    }

    fun submit() {
        val st = _state.value
        if (st.isSubmitting || st.quizzes.isEmpty()) return
        _state.update { it.copy(isSubmitting = true) }

        viewModelScope.launch {
            readingLogId?.let { logId ->
                try {
                    ApiClient.patch<ReadingLogResponse>(
                        path = "/api/logs",
                        body = PatchLogRequest(
                            id = logId,
                            progressRatio = null,
                            finishedAtUnix = System.currentTimeMillis() / 1000,
                            quizScore = st.score,
                        ),
                    )
                } catch (_: Throwable) {
                    // 점수 저장 실패해도 결과 화면은 보여준다.
                }
            }
            _state.update { it.copy(isSubmitting = false, phase = Phase.Finished) }
        }
    }

    fun restart() {
        _state.update { it.copy(selections = emptyMap(), currentIndex = 0, phase = Phase.Answering) }
    }
}

@Serializable
internal data class PatchLogRequest(
    val id: Int,
    val progressRatio: Double? = null,
    val finishedAtUnix: Long? = null,
    val quizScore: Int? = null,
)

File: apps/ios/HaruBook/Features/Quiz/QuizViewModel.swift
import Foundation
import Observation

@Observable
@MainActor
final class QuizViewModel {
    enum Phase: Equatable {
        case loading
        case answering
        case finished
        case error(String)
    }

    let book: Book
    /// Reader에서 시작된 reading_log id. 점수 PATCH 대상. nil 이면 기록 생략.
    let readingLogId: Int?

    private(set) var quizzes: [Quiz] = []
    /// `quiz.id → 사용자가 선택한 choice index`.
    private(set) var selections: [Int: Int] = [:]
    private(set) var currentIndex: Int = 0
    private(set) var phase: Phase = .loading
    private(set) var isSubmitting: Bool = false

    init(book: Book, readingLogId: Int?) {
        self.book = book
        self.readingLogId = readingLogId
    }

    var currentQuiz: Quiz? { quizzes.indices.contains(currentIndex) ? quizzes[currentIndex] : nil }
    var totalQuestions: Int { quizzes.count }
    var isLastQuestion: Bool { currentIndex + 1 >= quizzes.count }
    var score: Int {
        quizzes.reduce(0) { acc, quiz in
            acc + (selections[quiz.id] == quiz.answerIndex ? 1 : 0)
        }
    }

    func load() async {
        phase = .loading
        do {
            let response: QuizzesResponse = try await APIClient.shared.send(
                Endpoint(path: "/api/books/\(book.id)/quiz", method: .post)
            )
            // `orderIndex` 오름차순으로 정렬.
            quizzes = response.quizzes.sorted(by: { $0.orderIndex < $1.orderIndex })
            phase = quizzes.isEmpty ? .error("퀴즈가 아직 준비되지 않았어요.") : .answering
        } catch {
            // QuizView가 자체 헤더 없이 메시지만 노출하므로 친화 문구 그대로.
            phase = .error(error.localizedDescription)
        }
    }

    func selectAnswer(_ index: Int) {
        guard let quiz = currentQuiz else { return }
        selections[quiz.id] = index
    }

    func goToNext() {
        guard currentIndex + 1 < quizzes.count else { return }
        currentIndex += 1
    }

    func goToPrevious() {
        guard currentIndex > 0 else { return }
        currentIndex -= 1
    }

    func submit() async {
        guard !isSubmitting, !quizzes.isEmpty else { return }
        isSubmitting = true
        defer { isSubmitting = false }

        // 점수 PATCH (logId가 있을 때만)
        if let logId = readingLogId {
            do {
                let _: ReadingLogResponse = try await APIClient.shared.send(
                    Endpoint(
                        path: "/api/logs",
                        method: .patch,
                        body: PatchLogRequest(
                            id: logId,
                            progressRatio: nil,
                            finishedAtUnix: Int(Date().timeIntervalSince1970),
                            quizScore: score
                        )
                    )
                )
            } catch {
                // 점수 저장 실패해도 결과 화면은 보여준다 (소프트 페일).
            }
        }

        phase = .finished
    }

    func restart() {
        selections.removeAll()
        currentIndex = 0
        phase = .answering
    }
}

/// `/api/logs` PATCH 본문 — Reader와 별도로 정의해 의존성을 끊는다.
private struct PatchLogRequest: Encodable {
    let id: Int
    let progressRatio: Double?
    let finishedAtUnix: Int?
    let quizScore: Int?
}
